NDTensors/ext/NDTensorsCUDAExt/append.jl

```diff
@@ -0,0 +1,9 @@
+## CUDA.jl has no append! function and 
+## calling base results in a scalar issue
+## This is probably super slow and I am seeing 
+## very slow blocksparse DMRG but it does work.
+function Base.append!(a::CuArray, b::CuArray)
```

Comment:
I made this append function but most likely this is type piracy and we need to talk to CUDA.jl about creating this function. For now this is a quick hack to get DMRG working with no scalar indexing error. Additionally we could just use `@allowscalar` when we have append. Or convert to CPU.

Reply:
I see, I guess `append!` also has a scalar indexing issue for `MtlArray`:
```julia
julia> using Metal

julia> v = mtl(randn(4))
4-element MtlVector{Float32, Private}:
  0.9403535
  0.89847064
  1.4950696
 -0.33930042

julia> Metal.@allowscalar append!(v, [2, 3])
6-element MtlVector{Float32, Private}:
  0.9403535
  0.89847064
  1.4950696
 -0.33930042
  2.0
  3.0

julia> v
6-element MtlVector{Float32, Private}:
  0.9403535
  0.89847064
  1.4950696
 -0.33930042
  2.0
  3.0

julia> append!(v, [2, 3])
ERROR: Scalar indexing is disallowed.
Invocation of setindex! resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore should be avoided.

If you want to allow scalar iteration, use `allowscalar` or `@allowscalar`
to enable scalar iteration globally or for the operations in question.
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] errorscalar(op::String)
   @ GPUArraysCore ~/.julia/packages/GPUArraysCore/GMsgk/src/GPUArraysCore.jl:155
 [3] _assertscalar(op::String, behavior::GPUArraysCore.ScalarIndexing)
   @ GPUArraysCore ~/.julia/packages/GPUArraysCore/GMsgk/src/GPUArraysCore.jl:128
 [4] assertscalar(op::String)
   @ GPUArraysCore ~/.julia/packages/GPUArraysCore/GMsgk/src/GPUArraysCore.jl:116
 [5] setindex!
   @ ~/.julia/packages/GPUArrays/Hd5Sk/src/host/indexing.jl:56 [inlined]
 [6] _append!(a::MtlVector{Float32, Private}, ::Base.HasShape{1}, iter::Vector{Int64})
   @ Base ./array.jl:1201
 [7] append!(a::MtlVector{Float32, Private}, iter::Vector{Int64})
   @ Base ./array.jl:1187
 [8] top-level scope
   @ REPL[5]:1
 [9] top-level scope
   @ ~/.julia/dev/Metal/src/initialization.jl:57
```

Reply:
A slightly better implementation could be:
```julia
function Base.append!(a::CuVector, items::AbstractVector)
  n = length(items)
  resize!(a, length(a) + n)
  copyto!(a, length(a) - n + 1, items, firstindex(items), n)
  return a
end
```
which is closer to the definition of `copyto!` for `Vector` in `Base`: https://github.com/JuliaLang/julia/blob/v1.10.2/base/array.jl#L1179-L1185. Note that version can also handle appending a CPU vector onto a GPU vector, as long as `copyto!` can handle that, which it looks like it does for CUDA.jl (https://github.com/JuliaGPU/CUDA.jl/blob/v5.2.0/src/array.jl#L534-L574) and Metal.jl (https://github.com/JuliaGPU/Metal.jl/blob/main/src/array.jl#L267-L312).

I suppose as an optimization that `Base.unsage_copyto!` could be used in exchange for `copyto!`, as you did, since we are sure that the inputs of `copyto!` are not out of bounds.

Reply:
I would say for the sake of this PR, let's not do this kind of type piracy with `append!`, and instead call `append!` with `@allowscalar`. Then we can make PRs to the various GPU backends implementing `append!` as a followup.

Reply:
For this issue, I made an expose version of `Base.append!` and for the generic version I call `append!` but for `Type{<:AbstractGPUArray}` I use `@allowscalar` to work generically across GPU backends. This might be too strongly opinionated so I can push the function to each specific backend if that is a better idea. Thanks!

Reply:
I made `Base.append!` functions for each GPU backend in their extension folders. I am closing this comment